fix: TypeError when loading CJS files after bundling mixed CJS+TS project

IWANABETHATGUY opened the pull request:
closed #5930

Current issue: 
https://repl.rolldown.rs/#eNplUctOwzAQ/JWVT6mIXHHEqEeOiANHzMFNNsGVY6d+FKQq/87aTlFET/a+Zmdmr2xg4sompS0/hfy1TPyFLeso0tPsfITJ9TB4N4FkfO+TQcmepV2LKgSkZ61b16OoqdIkbedsiDCbNGoLB7hKCxmwlXahem3lIXrdxZdzUqaprZx6+IRRtfD4tKNO/CnrehxUMjdAyhNVZCL6hEvLjD5uxKxR1UJ4xJxXmJCZ5P2MhrKizdQtrGOVfnbgAB7PSXtssg2ELVnmJe09cta438PH2/GEXQRXns+inBQJ8qm6yA3aMX7BA7yTAXZsCGq3JskggM6jiihgSLaL2llodhWdSkTM0VrjRiJknOqxhwqbaQEsq8VFojOmd9+W09Sgx63a+0oV/s/vstWlOKcobhQG5ydFoWTdKUhWGAPMHummF3wttgQB+TaFEPFhC7G55NuQrhDZ8gst29p6

The binding finalization only considers when the required CJS module is in a common chunk, 
so it always converts 
```
const mod = require('./lib')
```
to 
```js
const mod = require_rule$1.require_rule()
```

This did not consider if the wrap_ref is exported via entry cjs module, which may generate 
```
Object.defineProperty(exports, 'default', {
  enumerable: true,
  get: function () {
    return require_rule();
  }
});
```

this pr fixed this scenario.

IWANABETHATGUY commented:
* **#6755** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/6755?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6743** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/6743?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/6743?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add the label <kbd>graphite: merge</kbd> to this PR to add it to the merge queue.

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/rolldown?ref=merge-queue-instructions-in-stack-comment&prId=6743).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

hyf0 commented on crates/rolldown/src/module_finalizers/mod.rs:
```diff
@@ -277,13 +296,21 @@ impl<'me, 'ast> ScopeHoistingFinalizer<'me, 'ast> {
+              self.snippet.literal_prop_access_member_expr_expr(require_binding, "default")
+            } else {
+              let exported_name = &self.ctx.chunk_graph.chunk_table[chunk_idx_of_canonical_symbol]
+                .exports_to_other_chunks[&canonical_ref][0];
```
Would it be more proper to prevent the wrong logic in the begaining instead of catching it up here?

`&self.ctx.chunk_graph.chunk_table[chunk_idx_of_canonical_symbol]
                .exports_to_other_chunks[&canonical_ref][0]` seems just produce a wrong result based on the repl https://repl.rolldown.rs/#eNplUctOwzAQ/JWVT6mIXHHEqEeOiANHzMFNNsGVY6d+FKQq/87aTlFET/a+Zmdmr2xg4sompS0/hfy1TPyFLeso0tPsfITJ9TB4N4FkfO+TQcmepV2LKgSkZ61b16OoqdIkbedsiDCbNGoLB7hKCxmwlXahem3lIXrdxZdzUqaprZx6+IRRtfD4tKNO/CnrehxUMjdAyhNVZCL6hEvLjD5uxKxR1UJ4xJxXmJCZ5P2MhrKizdQtrGOVfnbgAB7PSXtssg2ELVnmJe09cta438PH2/GEXQRXns+inBQJ8qm6yA3aMX7BA7yTAXZsCGq3JskggM6jiihgSLaL2llodhWdSkTM0VrjRiJknOqxhwqbaQEsq8VFojOmd9+W09Sgx63a+0oV/s/vstWlOKcobhQG5ydFoWTdKUhWGAPMHummF3wttgQB+TaFEPFhC7G55NuQrhDZ8gst29p6

IWANABETHATGUY replied:
The `default` binding is generated at the chunk level by string replace https://github.com/rolldown/rolldown/blob/8406bc782847411c199d321034fe5f7475667dba/crates/rolldown/src/utils/chunk/render_chunk_exports.rs?plain=1#L266-L280 .  .To do that, we need to bind the `canonical_ref` to chunk-level default binding; that's not possible for now.

IWANABETHATGUY replied:
Wait, it is expected to do that here, since the wrapper_ref binding generation depends on whether the `wrapper_ref` declaration is in the same chunk as the importer

IWANABETHATGUY committed:
fix: TypeError when loading CJS files after bundling mixed CJS+TS project  (#6743)

closed #5930

Current issue:
https://repl.rolldown.rs/#eNplUctOwzAQ/JWVT6mIXHHEqEeOiANHzMFNNsGVY6d+FKQq/87aTlFET/a+Zmdmr2xg4sompS0/hfy1TPyFLeso0tPsfITJ9TB4N4FkfO+TQcmepV2LKgSkZ61b16OoqdIkbedsiDCbNGoLB7hKCxmwlXahem3lIXrdxZdzUqaprZx6+IRRtfD4tKNO/CnrehxUMjdAyhNVZCL6hEvLjD5uxKxR1UJ4xJxXmJCZ5P2MhrKizdQtrGOVfnbgAB7PSXtssg2ELVnmJe09cta438PH2/GEXQRXns+inBQJ8qm6yA3aMX7BA7yTAXZsCGq3JskggM6jiihgSLaL2llodhWdSkTM0VrjRiJknOqxhwqbaQEsq8VFojOmd9+W09Sgx63a+0oV/s/vstWlOKcobhQG5ydFoWTdKUhWGAPMHummF3wttgQB+TaFEPFhC7G55NuQrhDZ8gst29p6

The binding finalization only considers when the required CJS module is in a common chunk,
so it always converts
```
const mod = require('./lib')
```
to
```js
const mod = require_rule$1.require_rule()
```

This did not consider if the wrap_ref is exported via entry cjs module, which may generate
```
Object.defineProperty(exports, 'default', {
  enumerable: true,
  get: function () {
    return require_rule();
  }
});
```

this pr fixed this scenario.
